### PR TITLE
Rework virtual function page in ProgrammingGuide

### PR DESCRIPTION
**Pull request by JBludau**
In the wake of #529 and #530 some inconsistencies in this page became apparent. 
Thus this PR proposes:

- Renaming the classes in the example `Base` and `Derived` and the virtual function `Bar` as in the corresponding images.
- changing the wording when describing a call to `new` from `initializing` to `constructing`.
- Filling in some text for readability.
- corrected the problems with the images identified in #530. -> redrawn with fixes to the stated issues and linking them to the text

### CODE REVIEW

**Comment by JBludau** [comment]
Nice, this time git seems to show the pngs in the 'files changed' view

**Comment by ajpowelsnl** [comment]
Hi @JBludau - Many thanks for submitting this PR.  I **do** hate to rain on your parade, but for now, we are 110% focused on revising the [API guide](https://kokkos.org/kokkos-core-wiki/), and not the Programming guide.  Our goals are (in loose order):  1) fill in any gaps, 2) make sure documentation is correct with respect to the latest release, 3) correct errors, 4) improve / extend examples.  If you would like help getting up and running, please LMK, and I'll be happy to help.  Sorry to be a PR party pooper.

File: docs/source/ProgrammingGuide/Kokkos-and-Virtual-Functions.md [hunk 1]
```diff
@@ -45,44 +45,53 @@ Credit: the content of this section is adapted from Pablo Arias [here](https://p
...
-Instance* deviceInstance; // cudaMalloc'd to allocate it
+Base* deviceInstance; // cudaMalloc'd to allocate it
 Kokkos::parallel_for(... {
-  new((Instance*)deviceInstance) Instance(); // initialize an instance, and place the result in the pointer deviceInstance
+  new((Derived*)deviceInstance) Derived(); // construct an instance in the place, the pointer deviceInstance points to
 });
 ```
 
-This code is extremely ugly, but leads to a properly initialized instance of the class. Note that like with other uses of `new`, you need to later `free` the memory
+This code is extremely ugly, but leads to functional virtual function calls on the device. The vpointer now points to the device Vtable.
```
Comment (ldh4) [inline docs/source/ProgrammingGuide/Kokkos-and-Virtual-Functions.md:80]: Looks like there are mixed forms of capitalizations used for `vtable` in these changes (`vtable`/`Vtable`/`VTable`). Just for aesthetics, I suggest making them consistent.

Similar issue with `vpointer`.

Comment (JBludau) [reply docs/source/ProgrammingGuide/Kokkos-and-Virtual-Functions.md]: done. It should now be `Vpointer` and `Vtable` everywhere.

**Review by crtrott — APPROVED** [review]

**Pull request merged**